### ISSUE TITLE
Explainer: attn_merge_state (LSE-weighted partial-attention combine)

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#11]
## Summary

Reference explainer for **`attn_merge_state`** — the op that combines two partial attention outputs into one using their log-sum-exp (LSE) values. It is the glue behind split-KV decoding ("flash decoding") and chunked/streamed prefix attention. Informational.

Code: `tokenspeed-kernel/.../ops/attention/` (public API `__init__.py`, Triton kernel `triton/__init__.py`, CUDA variant `cuda/__init__.py`).

## What it does

Softmax attention can be computed over disjoint KV slices independently, each producing a partial output `O_i` and partial `LSE_i = log Σ exp(scores)`. `attn_merge_state` fuses `(O_a, LSE_a)` and `(O_b, LSE_b)` into the exact union result without recomputation:

```
m = max(LSE_a, LSE_b);  w_a = e^(LSE_a-m), w_b = e^(LSE_b-m)
O = (O_a*w_a + O_b*w_b)/(w_a+w_b);  LSE = m + log(w_a+w_b)
```

It emits the merged LSE so merges **chain**, and is **producer-agnostic** (only consumes outputs + LSEs) → one impl covers MHA and MLA.

## Implementation notes

- Triton kernel works in base-2 (`exp2`/`log2`); incoming natural-log LSE scaled by `lse_scale_log2` (default `log2(e)`; pass `1.0` if producer already emits log2).
- All math promoted to fp32 internally; output stored in input dtype. Grid = `total_q * num_heads` (element-parallel, cheap). `BLOCK_D = next_power_of_2(head_dim)` with mask.
- Backends: `triton_attn_merge_state` (PORTABLE default) + `cuda_attn_merge_state` (specialized, supports in-place accumulate).

## Where it's used

- **Split-KV decode:** the Triton decode kernel's stage-2 reduction inlines the identical N-way LSE-weighted merge (partials live in scratch `attn_logits`/`attn_lse`), so decode doesn't call the standalone op.
- **MLA chunked prefix (DeepSeek):** each prefix KV chunk's attention is merged in-place into a running accumulator (`merge_state(..., inplace=True)`).
- **MHA extend-with-kvcache:** `return_lse=True` exposes per-row LSE so an outer step can combine extend with a cache contribution.

## Gotchas

Exact & order-independent; needs LSE (not just outputs); fp32 internal accumulation; keep `lse_scale_log2` matched to the producer domain (mismatch silently skews weights); in-place CUDA variant avoids per-chunk allocations in long loops.

---
_Filed from local engineering notes (`~/agentic_results/merge_attn_state_tokenspeed.md`). Informational._
